pkg/controllers/clusterresourcesnapshot/watcher.go

```diff
@@ -0,0 +1,105 @@
+	RolloutController controller.Controller
+}
+
+// Reconcile triggers a single CRP reconcile round when scheduling policy has changed.
+func (r *Watcher) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	name := req.NamespacedName
+	snapshot := fleetv1beta1.ClusterResourceSnapshot{}
+	snapshotKRef := klog.KRef(name.Namespace, name.Name)
+
+	startTime := time.Now()
+	klog.V(2).InfoS("Reconciliation starts", "clusterResourceSnapshot", snapshotKRef)
+	defer func() {
+		latency := time.Since(startTime).Milliseconds()
+		klog.V(2).InfoS("Reconciliation ends", "clusterResourceSnapshot", snapshotKRef, "latency", latency)
+	}()
+
+	if err := r.Client.Get(ctx, name, &snapshot); err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(4).InfoS("Ignoring NotFound clusterResourceSnapshot", "clusterResourceSnapshot", snapshotKRef)
+			return ctrl.Result{}, nil
+		}
+		klog.ErrorS(err, "Failed to get clusterResourceSnapshot", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, controller.NewAPIServerError(true, err)
+	}
+	// check if it is the first resource snapshot
+	_, exist := snapshot.Annotations[fleetv1beta1.NumberOfResourceSnapshotsAnnotation]
+	if !exist {
+		// we only care about when a new resource snapshot index is created
+		klog.V(2).InfoS("Ignore the subsequent sub resource snapshots", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, nil
+	}
+	// check if it is the latest resource snapshot
+	isLatest, err := strconv.ParseBool(snapshot.Labels[fleetv1beta1.IsLatestSnapshotLabel])
+	if err != nil {
+		err := fmt.Errorf("invalid annotation value %s", fleetv1beta1.IsLatestSnapshotLabel)
+		klog.ErrorS(err, "resource snapshot has does not have valid is latest annotation", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, controller.NewUnexpectedBehaviorError(err)
+	}
+	if !isLatest {
+		// All newly created resource snapshots should start with the latest label to be true.
+		// However, this can happen if the label is removed fast by the time this reconcile loop is triggered.
+		klog.V(2).InfoS("Newly created resource snapshot %s is not the latest", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, nil
+	}
+	// get the CRP name from the label
+	crp := snapshot.Labels[fleetv1beta1.CRPTrackingLabel]
+	if len(crp) == 0 {
+		err := fmt.Errorf("invalid CRPTrackingLabel label value %s", crp)
+		klog.ErrorS(err, "Invalid clusterResourceSnapshot", "clusterResourceSnapshot", snapshotKRef)
+		return ctrl.Result{}, controller.NewUnexpectedBehaviorError(err)
+	}
+	// enqueue the CRP to the rollout controller queue
+	r.RolloutController.Enqueue(crp)
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *Watcher) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&fleetv1beta1.ClusterResourceSnapshot{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
```

Comment:
Hi Ryan! This will lead the controller to ignore metadata changes; if the CRP controller adds a label/annotation, the event will not be picked up.

Reply:
The resource snapshot spec is immutable (ideally), so this predicate will filter out every change I guess.

Reply:
yeah, is there any problem with filter out any metadata/status changes?